appliance/redis/process.go

```diff
@@ -0,0 +1,439 @@
+	ErrStopped = errors.New("redis already stopped")
+)
+
+// Process represents a running Redis process.
+type Process struct {
+	mtx     sync.Mutex
+	running bool
+
+	stopping atomic.Value
+	stopped  chan struct{}
+
+	// The daemon running redis-server.
+	cmd *exec.Cmd
+
+	ID           string
+	Singleton    bool
+	Port         string
+	BinDir       string
+	DataDir      string
+	Password     string
+	OpTimeout    time.Duration
+	ReplTimeout  time.Duration
+	Logger       log15.Logger
+	WaitUpstream bool
+}
+
+// NewProcess returns a new instance of Process with defaults.
+func NewProcess() *Process {
+	p := &Process{
+		Port:        DefaultPort,
+		BinDir:      DefaultBinDir,
+		DataDir:     DefaultDataDir,
+		Password:    DefaultPassword,
+		OpTimeout:   DefaultOpTimeout,
+		ReplTimeout: DefaultReplTimeout,
+		Logger:      log15.New("app", "redis"),
+	}
+	p.stopping.Store(false)
+	return p
+}
+
+// ConfigPath returns the path to the redis.conf.
+func (p *Process) ConfigPath() string { return filepath.Join(p.DataDir, "redis.conf") }
+
+// Start begins the process.
+// Returns an error if the process is already running.
+func (p *Process) Start() error {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	// Valdiate that process is not already running and that we have a config.
+	if p.running {
+		return ErrRunning
+	}
+	return p.start()
+}
+
+func (p *Process) start() error {
+	logger := p.Logger.New("fn", "start", "data_dir", p.DataDir, "bin_dir", p.BinDir)
+
+	// Initialize stop state so we know if it's an expected stop.
+	p.stopping.Store(false)
+	p.stopped = make(chan struct{})
+
+	// Generate configuration.
+	if err := p.writeConfig(); err != nil {
+		logger.Error("error writing conf", "path", p.ConfigPath(), "err", err)
+		return err
+	}
+
+	logger.Info("starting process")
+
+	// Execute redis-server binary.
+	cmd := exec.Command(filepath.Join(p.BinDir, "redis-server"), p.ConfigPath())
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	if err := cmd.Start(); err != nil {
+		logger.Error("failed to start", "err", err)
+		return err
+	}
+	p.cmd = cmd
+	p.running = true
+
+	logger.Info("process started")
+
+	// Monitor redis process to check for unexpected quit.
+	go p.monitorCmd(p.cmd, p.stopped)
+
+	// Ping until successful response.
+	if err := p.pingWait("", p.OpTimeout); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Stop attempts to gracefully stop the process. If the process cannot be
+// stopped gracefully then it is forcefully stopped. Returns an error if the
+// process is already stopped.
+func (p *Process) Stop() error {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	if !p.running {
+		return ErrStopped
+	}
+	return p.stop()
+}
+
+func (p *Process) stop() error {
+	logger := p.Logger.New("fn", "stop")
+	logger.Info("stopping")
+
+	// Mark process as expecting a shutdown.
+	p.stopping.Store(true)
+
+	// Attempt to kill via escalating signals.
+	for _, sig := range []os.Signal{syscall.SIGINT, syscall.SIGQUIT, syscall.SIGKILL} {
```

Comment:
These were the incorrect signals. Changed to `SIGTERM` (graceful shutdown) and `SIGSEV` (hard shutdown) in c6b8b6a.

http://redis.io/topics/signals